### PR TITLE
#134249131 Users should receive a JWT on successful signup

### PR DESCRIPTION
**Pull request by andela-sayidu**
#### What does this PR do?             
This PR ensure that once a user successfully signs up,  they are assigned a JSON token  which is used for subsequent calls to all other endpoints that require authentication.
             
#### How should this be manually tested?     
This PR can be tested manually by running a local server and using POSTMAN to verify that the user connections are addressed with appropriate responses and JWT Tokens are assigned upon successfully signup. 
                            
#### Any background context you want to provide?                           
Story 134249129 : Users should receive a JWT on successful login. 
Will support the usage of JWT Tokens upon login.

### CODE REVIEW

**Comment by adebayo-maborukoje** [comment]
could you correct this statement?  `successfully signs up successfully` on the PR description.

File: .gitignore [hunk 1]
```diff
@@ -6,3 +6,6 @@ c4h_keys.txt
 .env
 jasmine
 coverage
+coverage/
+.gitignore
```
Comment (adebayo-maborukoje) [inline .gitignore]: why are you ignoring gitignore in .gitignore?

File: app/controllers/auth.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+              });
+            }
+          } else {
+            res.json({
+              message: 'User already exists!'
+            });
+          }
+        });
+    };
+
```
Comment (adebayo-maborukoje) [inline app/controllers/auth.js]: Maintain a indentation consistency in this file. I see you using 4spaces (TAB) in some lines and uses 2 spaces (Space) in other parts.

File: app/controllers/auth.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
+'use strict';
+
+let mongoose = require('mongoose');
```
Comment (adebayo-maborukoje) [inline app/controllers/auth.js]: change all this to `const` and not `let`.

File: app/controllers/auth.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            user.email = email;
+            user.password = pwd;
+            user.avatar = avatar;
+
+            user.save((err) => {
+                  if (err) {
+                    res.send(err);
+                  }
+                  let token = jwt.sign({
+                    registeredUser,
```
Comment (adebayo-maborukoje) [inline app/controllers/auth.js]: What is `registeredUser`  in this place?

File: app/controllers/auth.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                    expiresInMinutes: 1440
```
Comment (adebayo-maborukoje) [inline app/controllers/auth.js]: can we not hardcode the expiration time here?

File: app/controllers/auth.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        if (err) {
+          res.send(err);
+        }
+
+        if (!registeredUser) {
+          if (name && email && pwd) {
+            let user = new User();
+            user.name = name;
+            user.email = email;
+            user.password = pwd;
```
Comment (adebayo-maborukoje) [inline app/controllers/auth.js]: I think this password needs some form of validation. else i can just `space` and create a password.

File: app/controllers/auth.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+            user.save((err) => {
+                  if (err) {
+                    res.send(err);
+                  }
+                  let token = jwt.sign({
+                    registeredUser,
+                    expiresInMinutes: 1440
+                  }, config.secret);
+                  res.json({
```
Comment (adebayo-maborukoje) [inline app/controllers/auth.js]: What statusCode are you sending here?

File: server.js [hunk 8]
```diff
@@ -8,6 +8,9 @@ const express = require('express'),
     logger = require('mean-logger'),
     io = require('socket.io');
     require('dotenv').config({silent: true});
+const session = require('express-session');
```
Comment (adebayo-maborukoje) [inline server.js]: maintain the same declaration style in this file.

File: test/client/signup-test.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  });
+
+
+  it('verifies that a user can not register more than sucessful', (done) => {
+    let user = {
+      name: 'Andela',
+      email: 'andela@yahoo.com',
+      password: '12345',
+      avatar: 'andela1'
+    };
```
Comment (adebayo-maborukoje) [inline test/client/signup-test.js]: Add strict mode to this file as suggested by HoundCI

File: test/client/signup-test.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        expect(res.body.message).to.equal('Thank you for signing up!');
+        expect(typeof res.body.token).to.equal('string');
+        expect(res).to.have.header('content-type', 'application/json; charset=utf-8');
+        expect(res).to.have.status(200);
+        done();
+      });
+  });
+
+
+  it('verifies that a user can not register more than sucessful', (done) => {
```
Comment (adebayo-maborukoje) [inline test/client/signup-test.js]: test description is not clear enough.

File: test/client/signup-test.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+      avatar: 'andela1'
+    };
+
+    chai.request(server)
+      .post('/api/auth/signup')
+      .send(user)
+      .end((err, res) => {
+        expect(res.body.message).to.equal('User already exists!');
+        expect(res).to.have.header('content-type', 'application/json; charset=utf-8');
+        expect(res).to.have.status(200);
```
Comment (adebayo-maborukoje) [inline test/client/signup-test.js]: why should an unsuccessful signup have a 200 status.

File: .gitignore (hunk 1 above)
Comment (andela-sayidu) [inline .gitignore]: I have addressed this, was trying this out locally.

File: test/client/signup-test.js (hunk 9 above)
Comment (andela-sayidu) [inline test/client/signup-test.js]: Addressed

File: server.js [hunk 12]
```diff
@@ -17,18 +22,19 @@ const express = require('express'),
 //if test env, load example file
 
 process.env.NODE_ENV = process.env.NODE_ENV || 'development';
-const config = require('./config/config'),
-      auth = require('./config/middlewares/authorization'),
-      mongoose = require('mongoose');
+console.log(process.env.NODE_ENV);
```
Comment (andela-sayidu) [inline server.js]: Nit-Pick 

Remove unrequired console statement

File: app/controllers/auth.js [hunk 13]
```diff
@@ -1,15 +1,79 @@
...
-const config = require('../../config/env/all');
-const moment = require('moment');
+const jwt = require('jsonwebtoken');
+const moments = require('moment');
+const config = require('../../config/config');
+const validate = require('validator');
 const User = mongoose.model('User');
 
+exports.signUp = (req, res) => {
+  console.log(req.body);
```
Comment (adebayo-maborukoje) [inline app/controllers/auth.js]: *NIT PICK*
remove Console statement

Comment (andela-sayidu) [inline app/controllers/auth.js]: Addressed

**Pull request merged**